lib/core/dns-resolver.js

```diff
@@ -0,0 +1,443 @@
+    this.stats = {
+      cache: 0,
+      query: 0
+    }
+
+    if (this._resolver instanceof AsyncResolver) {
+      this._resolve4 = this._resolver.resolve4.bind(this._resolver)
+      this._resolve6 = this._resolver.resolve6.bind(this._resolver)
+    } else {
+      this._resolve4 = promisify(this._resolver.resolve4.bind(this._resolver))
+      this._resolve6 = promisify(this._resolver.resolve6.bind(this._resolver))
+    }
+
+    this._iface = getIfaceInfo()
+
+    this._pending = {}
+    this._nextRemovalTime = false
+    this._hostnamesToFallback = new Set()
+
+    this.fallbackDuration = fallbackDuration
+
+    if (fallbackDuration > 0) {
+      const interval = setInterval(() => {
+        this._hostnamesToFallback.clear()
+      }, fallbackDuration * 1000)
+
+      /* istanbul ignore next: There is no `interval.unref()` when running inside an Electron renderer */
+      if (interval.unref) {
+        interval.unref()
+      }
+
+      this._fallbackInterval = interval
+    }
+
+    if (lookupOptions) {
+      this.lookup = (hostname, _options, callback) => this.__lookup(hostname, lookupOptions, callback)
+    } else {
+      this.lookup = this.__lookup.bind(this)
+    }
+    this.lookupAsync = this.lookupAsync.bind(this)
+  }
+
+  set servers (servers) {
+    this.clear()
+
+    this._resolver.setServers(servers)
+  }
+
+  get servers () {
+    return this._resolver.getServers()
+  }
+
+  __lookup (hostname, options, callback) {
+    if (typeof options === 'function') {
+      callback = options
+      options = {}
+    } else if (typeof options === 'number') {
+      options = {
+        family: options
+      }
+    }
+
+    if (!callback) {
+      throw new Error('Callback must be a function.')
+    }
+
+    // eslint-disable-next-line promise/prefer-await-to-then
+    this.lookupAsync(hostname, options).then((result) => {
+      if (options.all) {
+        callback(null, result)
+      } else {
+        callback(
+          null,
+          result.address,
+          result.family,
+          result.expires,
+          result.ttl
+        )
+      }
+    }, callback)
+  }
+
+  async lookupAsync (hostname, options = {}) {
+    if (typeof options === 'number') {
+      options = {
+        family: options
+      }
+    }
+
+    let cached = await this.query(hostname)
+
+    if (options.family === 6) {
+      const filtered = cached.filter((entry) => entry.family === 6)
+
+      if (options.hints & V4MAPPED) {
+        if ((supportsALL && options.hints & ALL) || filtered.length === 0) {
+          map4to6(cached)
+        } else {
+          cached = filtered
+        }
+      } else {
+        cached = filtered
+      }
+    } else if (options.family === 4) {
+      cached = cached.filter((entry) => entry.family === 4)
+    }
+
+    if (options.hints & ADDRCONFIG) {
+      const { _iface } = this
+      cached = cached.filter((entry) =>
+        entry.family === 6 ? _iface.has6 : _iface.has4
+      )
```

Comment:
Please avoid the use of filter, map etc, they are an order of magnitude (or more) slower than using normal iteration.

Reply:
I translated the code with loops, no optimizations yet. To have the code passing all tests. Made the function less readable. Also note that here it will be common to have one or two items returns and I am yet to see anyone using more than 4-5 items per record. 
Agreeing to keep an eye on the perf side.